pallets/connectors/src/message.rs

```diff
@@ -76,244 +82,289 @@ impl<Domain: Encode, PoolId: Encode, TrancheId: Encode, Balance: Encode, Rate: E
 	}
 }
 
-impl<Domain: Encode, PoolId: Encode, TrancheId: Encode, Balance: Encode, Rate: Encode> Encode
-	for Message<Domain, PoolId, TrancheId, Balance, Rate>
+impl<
+		Domain: Encode + Decode,
+		PoolId: Encode + Decode,
+		TrancheId: Encode + Decode,
+		Balance: Encode + Decode,
+		Rate: Encode + Decode,
+	> EncodeLike for Message<Domain, PoolId, TrancheId, Balance, Rate>
+{
+}
+
+impl<
+		Domain: Encode + Decode,
+		PoolId: Encode + Decode,
+		TrancheId: Encode + Decode,
+		Balance: Encode + Decode,
+		Rate: Encode + Decode,
+	> Decode for Message<Domain, PoolId, TrancheId, Balance, Rate>
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let call_type = input.read_byte()?;
+
+		match call_type {
+			0 => Ok(Self::Invalid),
+			1 => Ok(Self::AddPool {
+				pool_id: decode_be_bytes::<8, _, _>(input)?,
```

Comment:
In case we dont like the custom primitive implementations of `MaxEncodedLen`, we can add our own trait:
```rust
trait EncondingSize {
    type Size: usize;
}
```

Reply:
Feel free to skip this if by now it's too cumberstone

Reply:
I really love the idea and suggestions here @lemunozm but if it's ok I will leave it as is for the reason that I like the idea of not mixing the number of bytes that a type encodes to for the Connector General Messaging format with a general approach like `MaxEncodedLen`. This also makes it very explicit to see how many bytes we read per field, just like we do in Solidity. Happy to be told otherwise and/or to revisit this later on 👍

Reply:
Mmm... But this pattern with enums appears a lot in all structs stored in pallets that contain enums (as `WriteOffStatus` of current `pallet-loans`). I understand that this value will be the `Max` value of all possible variants, wasting memory in the worst case.

Reply:
Not sure I follow?

Reply:
@mustermeiszer my point is that if we change this implementation to use a general-purpose trait like `MaxEncodedLen`, we are making use of a general purpose trait - that is used for other purposes like handling the affected types in storage - when the purpose of implementing that trait for those types is purely to serve the Connectors' General Message Passing Format.

We could still consider a custom trait like `EncondingSize` that @lemunozm suggested above, but I am just not sure if we benefit from introducing a trait, implementing it for all the types we need it for, only to not be explicit in a single place where it would be used (when decoding a Message).

This is my pov but yours (both of you ofc) is more than welcome so let me know what you think 🤟

Reply:
I wasn't referring to you, but rather to @lemunozm comment as I think using max encoded Len is not a good idea. 

I don't wanna use a custom trait, we can keep it as is.